crates/executor/src/blockchain_tree/mod.rs

```diff
@@ -0,0 +1,811 @@
+        }
+
+        let old_tip = self.block_indices.canonical_tip();
+        // Merge all chain into one chain.
+        let mut new_canon_chain = chains_to_promote.pop().expect("There is at least one block");
+        for chain in chains_to_promote.into_iter().rev() {
+            new_canon_chain.append_chain(chain);
+        }
+
+        // update canonical index
+        self.block_indices.canonicalize_blocks(&new_canon_chain.blocks);
+
+        // if joins to the tip
+        if new_canon_chain.fork_block_hash() == old_tip.hash {
+            // append to database
+            self.commit_canonical(new_canon_chain)?;
+        } else {
+            // it forks to canonical block that is not the tip.
+
+            let canon_fork = new_canon_chain.fork_block();
+            // sanity check
+            if self.block_indices.canonical_hash(&canon_fork.number) != Some(canon_fork.hash) {
+                unreachable!("all chains should point to canonical chain.");
+            }
+
+            // revert `N` blocks from current canonical chain and put them inside BlockchanTree
+            // This is main reorgs on tables.
+            let old_canon_chain = self.revert_canonical(canon_fork.number)?;
+            self.commit_canonical(new_canon_chain)?;
+
+            // insert old canonical chain to BlockchainTree.
+            // TODO check if there there is chains that can be merged.
+            self.insert_chain(old_canon_chain);
+        }
+
+        Ok(())
+    }
+
+    /// Commit chain for it to become canonical. Assume we are doing pending operation to db.
+    fn commit_canonical(&mut self, chain: Chain) -> Result<(), Error> {
+        let mut tx = Transaction::new(&self.externals.db)?;
+
+        let new_tip = chain.tip().number;
+
+        for item in chain.blocks.into_iter().zip(chain.changesets.into_iter()) {
+            let ((_, block), changeset) = item;
+            tx.insert_block(block, self.externals.chain_spec.as_ref(), changeset).map_err(|e| {
+                println!("commit error:{e:?}");
+                ExecError::VerificationFailed
+            })?;
+        }
+        // update pipeline progress.
+        tx.update_pipeline_stages(new_tip).map_err(|_| ExecError::VerificationFailed)?;
+
+        // TODO error cast
+
+        tx.commit()?;
+
+        Ok(())
+    }
+
+    /// Revert canonical blocks from database and insert them to pending table
+    /// Revert should be non inclusive, and revert_until should stay in db.
+    /// Return the chain that represent reverted canonical blocks.
+    fn revert_canonical(&mut self, revert_until: BlockNumber) -> Result<Chain, Error> {
+        // read data that is needed for new sidechain
+
+        let mut tx = Transaction::new(&self.externals.db)?;
+
+        // read block and execution result from database. and remove traces of block from tables.
+        let blocks_and_execution = tx
+            .get_block_and_execution_range::<true>(
```

Comment:
We should come up with a better name as `get_block_and_execution_range::<true>(..)` will actually also unwind the internal tables and the function name is not communicating that right now

Reply:
Having `take` and `get` would be better, will see to refactor it.